api/src/main/scala/types.scala

```diff
@@ -2294,8 +2294,8 @@ private[bson] sealed trait BSONDocumentLowPriority { self: BSONDocument =>
 
     lazy val fields = BSONDocument.toMap(elements)
 
-    val isEmpty = self.elements.isEmpty && producers.isEmpty
-    @inline def headOption = self.elements.headOption
+    def isEmpty = headOption.isEmpty
```

Comment:
`producers.isEmpty = false` is not reliable considering we have `Empty` `ElementProducer`

Reply:
side note, there's some compiler bug here, this ++ doesn't get called at all, we get `ClassCastException` instead.

I tried but couldn't form a reproducible version, have highlighted this as part of unit test, please refer [here](https://github.com/ReactiveMongo/ReactiveMongo-BSON/pull/208/files#diff-17dd49db27efa0de93dc88f8719aa8032f35c4979fdc1bf89880e3230ce29e5bR32).